### PR TITLE
[espnet3-4] Add support for multiple optimizers and schedulers

### PR DESCRIPTION
**Pull request by Masao-Someki**
## What did you change?

- Added new HybridOptim and HybridLRS classes under espnet3/trainer to enable unified management of multiple optimizers and schedulers.
- Enhanced LitESPnetModel to allow flexible configuration of multiple optimizers/schedulers, including strict validation:
    - Continued support for single optimizer/scheduler usage.
    - Each optimizer can be assigned parameters via the params key, with checks for overlaps, missing assignments, and required fields.
- Added comprehensive unit tests:
    - test/espnet3/test_hybrid_optim.py and test_hybrid_scheduler.py for HybridOptim and HybridLRS
    - test_model_with_optim_scheduler.py for various configuration and error scenarios (including both valid and invalid patterns).

---

## Why did you make this change?

- To support advanced training strategies such as per-module or partial training using multiple optimizers/schedulers.
- To avoid issues in PyTorch Lightning where training_step is called multiple times when using multiple optimizers.
- To improve maintainability and reliability by catching configuration and assignment errors early.

---

## Is your PR small enough?

Yes, the core content of this PR is focused and concise. Although it includes a significant number of additions, the code is well-documented and thoroughly tested.
Note that this PR depends on the following related PRs:

* #6139
* #6167
* #6171
* #6172

---

## Additional Context

- Core implementation is based on community discussion and recipes ([reference](https://github.com/Lightning-AI/pytorch-lightning/issues/3346#issuecomment-1478556073)), but with additional validation and safety checks.
- The test suite covers both normal and edge cases to ensure future extensibility and regression safety.
- #6133

### CODE REVIEW

**Comment by Masao-Someki** [comment]
## Test Case for HybridOptim

| Test Name                          | Description                                                              |
|-----------------------------------|--------------------------------------------------------------------------|
| test_zero_grad_called             | Ensures zero_grad() is called with correct arguments on all optimizers  |
| test_step_called_and_loss_returned| Confirms step() calls each optimizer's step and returns closure loss     |
| test_state_dict_and_load_state_dict_roundtrip | Tests saving and restoring of optimizer state dicts            |
| test_repr_contains_optimizer_info | Checks that __repr__ includes optimizer names (SGD, Adam, etc.)         |
| test_combined_properties          | Validates param_groups, state, defaults are aggregated correctly         |

**Comment by Masao-Someki** [comment]
## Test Case for HybridLRS

| Test Name                          | Description                                                              |
|-----------------------------------|--------------------------------------------------------------------------|
| test_hybrid_lrs_basic_attributes  | Validates optimizer, scheduler, and index assignment                    |
| test_getattr_forwarding           | Checks that method calls are correctly forwarded to lr_scheduler        |

Note that this test focuses on multiple schedulers and we didn't test ReduceOnPlateau, it is tested on `test_model_with_optim_scheduler.py`

**Comment by Masao-Someki** [comment]
## Test Case for LitESPnetModel.configure_optimizers


**Valid Configuration Tests**
| Test Name                           | Description                                                             |
|------------------------------------|-------------------------------------------------------------------------|
| test_single_optim_and_scheduler    | Validates single optimizer + scheduler configuration                    |
| test_multiple_optims_and_schedulers| Validates multiple optimizers and schedulers with param-based mapping   |
| test_custom_scheduler_interval     | Ensures scheduler interval is correctly set to "step"                   |
| test_reduce_on_plateau_with_config_adam | Tests ReduceLROnPlateau scheduler integration with manual stepping |

**Invalid Configuration Tests**
| Test Name                              | Description                                                             |
|---------------------------------------|-------------------------------------------------------------------------|
| test_missing_both_optim_and_optims    | Raises ValueError when neither `optim` nor `optims` is defined         |
| test_mixed_optim_and_optims           | Detects invalid mix of `optim` and `optims`                             |
| test_mixed_scheduler_and_schedulers   | Detects invalid mix of `scheduler` and `schedulers`                     |
| test_optims_and_schedulers_length_mismatch | Asserts that number of schedulers matches number of optimizers     |
| test_optimizer_missing_params_key     | Ensures `params` key is required for each optimizer config              |
| test_optimizer_params_not_matching_model | Asserts at least one parameter is matched for each `params` string   |
| test_optimizer_duplicate_params       | Fails if any parameter is assigned to more than one optimizer           |
| test_optimizer_missing_coverage       | Detects trainable parameters not assigned to any optimizer              |
| test_missing_nested_optim_block       | Ensures each `optims` block contains a nested `optim` definition        |

**Comment by sw005320** [comment]
@Emrys365, can you review this PR?

File: espnet3/trainer/hybrid_optim.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
+# This script is copied from
+# https://github.com/Lightning-AI/pytorch-lightning/issues/3346#issuecomment-1478556073
+
+from typing import Any, Callable, Dict, Iterable, List, Union
+
+import torch
+from typeguard import typechecked
+
+
+class HybridOptim(torch.optim.Optimizer):
```
Comment (sw005320) [inline espnet3/trainer/hybrid_optim.py]: `HybridOptim` sounds wired for me.
Why not `MultipleOptim`?
You actually use `multiple` in this PR name.

Comment (sw005320) [reply espnet3/trainer/hybrid_optim.py]: If this is based on some naming rules in the other modules (e.g., pytorch lightning), it is fine.

File: espnet3/trainer/hybrid_scheduler.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+from typing import Any, Dict, Iterable, List, Union
+
+import torch
+from typeguard import typechecked
+
+from espnet3.trainer.hybrid_optim import HybridOptim
+
+
+class HybridLRS(torch.optim.lr_scheduler._LRScheduler):
```
Comment (sw005320) [inline espnet3/trainer/hybrid_scheduler.py]: Ditto

File: espnet3/trainer/hybrid_optim.py (hunk 1 above)
Comment (sw005320) [reply espnet3/trainer/hybrid_optim.py]: Or `MultiOptim`

Comment (Masao-Someki) [reply espnet3/trainer/hybrid_optim.py]: This is just from the original script. I’ll change it to `MultipleOptim` (and `MultipleScheduler`), that sounds good!

**Comment by Emrys365** [comment]
> @Emrys365, can you review this PR?

Sure, I will review it.

File: espnet3/trainer/model.py [hunk 3]
```diff
@@ -183,15 +183,64 @@ def validation_step(self, batch, batch_idx):
...
+            - optim:
+                _target_: torch.optim.SGD
+                lr: 0.01
+                params: decoder
+
+        schedulers:
+            - _target_: torch.optim.lr_scheduler.StepLR
+              step_size: 10
+            - _target_: torch.optim.lr_scheduler.ReduceLROnPlateau
+              patience: 2
```
Comment (Emrys365) [inline espnet3/trainer/model.py]: Currently, the configuration list of `optims` requires each element to be a nested dictionary, while for `schedulers` each element is a flat dictionary.

How about unifying the formats?

File: espnet3/trainer/model.py [hunk 4]
```diff
@@ -213,44 +262,78 @@ def configure_optimizers(self):
...
+                getattr(self.config, "optim", None) is None
+            ), "Mixture of `optim` and `optims` is not allowed."
+            assert len(self.config.optims) == len(self.config.schedulers), (
+                f"The number of optimizers and schedulers must be equal: "
+                + f"optims: {len(self.config.optims)}, "
+                + f"schedulers: {len(self.config.schedulers)}"
+            )
+
+            optims = []
+            trainable_params = dict(self.named_parameters())  # key: name, value: param
```
Comment (Emrys365) [inline espnet3/trainer/model.py]: This is just a note that applying 'freeze_param' (freezing some parameters of the model) should be done before this line.

Comment (Emrys365) [reply espnet3/trainer/model.py]: BTW, it is better to apply the filter `if param.requires_grad` here so that you can shorten the code later.

```suggestion
            trainable_params = {
                name: param for name, param in self.named_parameters() if param.requires_grad
            }
```

File: espnet3/trainer/model.py [hunk 5]
```diff
@@ -213,44 +262,78 @@ def configure_optimizers(self):
...
+                selected = [
+                    p
+                    for name, p in trainable_params.items()
+                    if optim_cfg["params"] in name and p.requires_grad
+                ]
+                selected_names = [
+                    name
+                    for name, p in trainable_params.items()
+                    if optim_cfg["params"] in name and p.requires_grad
+                ]
```
Comment (Emrys365) [inline espnet3/trainer/model.py]:
```suggestion
                selected = [
                    p
                    for name, p in trainable_params.items()
                    if name in optim_cfg["params"]
                ]
                selected_names = [
                    name
                    for name, p in trainable_params.items()
                    if name in optim_cfg["params"]
                ]
```

File: espnet3/trainer/model.py [hunk 6]
```diff
@@ -6,12 +6,12 @@
...
 from espnet2.train.collate_fn import CommonCollateFn
+from espnet3.trainer.dataloader import DataLoaderBuilder
 
 # Temporarily disabled for the code review.
 # from espnet3.collect_stats import collect_stats
-# from espnet3.trainer.hybrid_optim import HybridOptim
-# from espnet3.trainer.hybrid_scheduler import HybridLRS
-from espnet3.trainer.dataloader import DataLoaderBuilder
+from espnet3.trainer.hybrid_optim import HybridOptim
+from espnet3.trainer.hybrid_scheduler import HybridLRS
```
Comment (Emrys365) [inline espnet3/trainer/model.py]: It seems that the `hybrid_optim` and `hybrid_scheduler` modules are not included in this PR.

File: espnet3/trainer/mutiple_scheduler.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    @typechecked
+    def __init__(
+        self,
+        multiple_optimizer: MultipleOptim,
+        lr_scheduler: torch.optim.lr_scheduler.LRScheduler,
+        optimizer_idx: int,
+    ) -> None:
+        self.optimizer = multiple_optimizer
+        self.lr_scheduler = lr_scheduler
+        self.idx = optimizer_idx
```
Comment (Emrys365) [inline espnet3/trainer/mutiple_scheduler.py]: Better to add a check to make sure 0 <= `optimizer_idx` < len(multiple_optimizer.optimizers)

File: test/espnet3/test_model_with_optim_scheduler.py [hunk 8]
```diff
@@ -0,0 +1,409 @@
...
+        self, epoch_nb, batch_nb, optimizer, optimizer_i, second_order_closure=None
+    ):
+        if batch_nb == 0:
+            self.scheduler.step(self.favorite_metric)
+            print(
+                f"metric: {self.favorite_metric}, best: {self.scheduler.best}, "
+                f"num_bad_epochs: {self.scheduler.num_bad_epochs}"
+            )
+        optimizer.step()
+        optimizer.zero_grad()
```
Comment (Emrys365) [inline test/espnet3/test_model_with_optim_scheduler.py]: `optimizer_i` and `second_order_closure` seem to be unused.

**Review by Emrys365 — APPROVED** [review]
The code looks good in general. I just left some minor comments.

File: espnet3/trainer/model.py (hunk 3 above)
Comment (Masao-Someki) [reply espnet3/trainer/model.py]: Sorry, the optim side should be the dict of optim and params.
It's tested in this test here:

https://github.com/Masao-Someki/espnet/blob/b685054e44449bf10989071de98c7e40fa622bdd/test/espnet3/test_model_with_optim_scheduler.py#L92-L121

Comment (Emrys365) [reply espnet3/trainer/model.py]: I see. Can you modify the scheduler config format similarly? Although it does not need other argiments for now, this might be more cleaer and better for future extensions.

**Comment by sw005320** [comment]
@Masao-Someki, please fix the name issues pointed out by copilot

**Review by Emrys365 — APPROVED** [review]

**Comment by sw005320** [comment]
Thanks!
